signals/signal_exporter.py

```diff
@@ -0,0 +1,252 @@
+class Signal:
+    publisher_id: str
+    process_name: str
+    event: str
+    runner_host: str = platform.node()
+    sample_no: int = -1
+    tag: str = "No tag specified"  # Could also be "user"
+    metadata: Dict = None
+
+    def __post_init__(self):
+        for (name, field_type) in self.__annotations__.items():
+            if not isinstance(self.__dict__[name], field_type):
+                if name == "metadata" and self.metadata == None:
+                    continue
+                raise TypeError(
+                    f"The field {name} should be type {field_type}, not {type(self.__dict__[name])}"
+                )
+
+    def to_json_str(self) -> Dict[str, Any]:
+        result: Dict[str, Any] = {
+            k: v
+            for k, v in self.__dict__.items()
+            if not (k.startswith("__") and k.endswith("__"))
+            and not (k == "metadata" and v == None)
+        }
+        return json.dumps(result)
+
+
+@dataclass
+class Response:
+    tool_id: str
+    publisher_id: str
+    event: str
+    ras: int
+
+    def __post_init__(self):
+        for (name, field_type) in self.__annotations__.items():
+            if not isinstance(self.__dict__[name], field_type):
+                if name == "ras" and self.ras == None:
+                    continue
+                raise TypeError(
+                    f"The field {name} should be type {field_type}, not {type(self.__dict__[name])}"
+                )
+
+    def to_json_str(self) -> Dict[str, Any]:
+        result: Dict[str, Any] = {
+            k: v
+            for k, v in self.__dict__.items()
+            if not (k.startswith("__") and k.endswith("__"))
+            and not (k == "ras" and v == None)
+        }
+        return json.dumps(result)
+
+
+class SignalExporter:
+    def __init__(
+        self, process_name, redis_host="localhost", redis_port=6379, runner_host=None
+    ) -> None:
+        # NOTE: runner_host will be automatically populated w/ platform.node() if nothing is passed in
+        self.subs = []
+        self.proc_name = process_name
+        self.runner_host = runner_host
+        self.pub_id = process_name + "-" + str(uuid.uuid4())
+        self.redis = redis.Redis(host=redis_host, port=redis_port, db=0)
+        self.init_listener = None
+        self.legal_events = None
+
+    def _sig_builder(self, event, sample=-1, tag=None, metadata=None):
+        sig = Signal(publisher_id=self.pub_id, process_name=self.proc_name, event=event)
+        sig.runner_host = self.runner_host if self.runner_host else sig.runner_host
+        sig.tag = tag if tag else sig.tag
+        sig.metadata = metadata if metadata else sig.metadata
+        sig.sample_no = sample if sample >= 0 else sig.sample_no
+        return sig
+
+    def _get_data_dict(self, response):
+        if not "data" in response:
+            print("No data in this response message")
+            return None
+        if isinstance(response["data"], int):
+            return None
```

Comment:
There was a case I ran into with initial connection/sub messages that would have an int data to report success, but the more general form you suggested is a much better way to go about it

Reply:
Oh wait actually it's not going to be a dict but a json string, so I can try/catch the json.loads() and return None on a Value error instead, and maybe also verify that its a string first to avoid relying solely on the try/catch (may hinder performance)